tests/test-heartbeat.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+/**
+ * @group heartbeat
+ */
+class Tests_Heartbeat extends WP_UnitTestCase {
+
+	function setUp() {
+		parent::setUp();
+		wp_clear_scheduled_hook( 'sp_heartbeat' );
+	}
+
+	function tearDown() {
+		wp_clear_scheduled_hook( 'sp_heartbeat' );
+		parent::tearDown();
+	}
+
+	function test_heartbeat_runs_automatically() {
+		$beat_result = SP_Heartbeat()->check_beat( true );
+		$this->assertTrue( $beat_result );
+		$this->assertTrue( wp_next_scheduled( 'sp_heartbeat' ) > 0 );
+
+		wp_clear_scheduled_hook( 'sp_heartbeat' );
+		$this->assertFalse( wp_next_scheduled( 'sp_heartbeat' ) );
+		wp_schedule_single_event( time(), 'sp_heartbeat', array( true ) );
+
+		// Run the cron
+		$pre_cron = time();
+		sp_tests_fake_cron();
+		$cron_time = time() - $pre_cron;
+
+		$this->assertTrue( wp_next_scheduled( 'sp_heartbeat' ) > 0 );
+
+		// The beat should have run and scheduled the next checkup. Here we make
+		// sure that the next event is scheduled in one heartbeat interval from
+		// now, while also accounting for however long the cron took.
+		$this->assertGreaterThanOrEqual(
+			time() + SP_Heartbeat()->intervals['heartbeat'] - $cron_time,
+			wp_next_scheduled( 'sp_heartbeat' )
+		);
+	}
+
+	function test_heartbeat_reschedules_after_query() {
+		wp_schedule_single_event( time(), 'sp_heartbeat' );
+		$next_scheduled = wp_next_scheduled( 'sp_heartbeat' );
+		SP_Heartbeat()->record_pulse();
+		$this->assertGreaterThan( $next_scheduled, wp_next_scheduled( 'sp_heartbeat' ) );
+	}
+
+	function test_heartbeat_increases_frequency_after_error() {
+		SP_Heartbeat()->record_pulse();
+		$was_scheduled = wp_next_scheduled( 'sp_heartbeat' );
+
+		// Change the host, re-check the heartbeat, and set the host back
+		$host = SP_Config()->get_setting( 'host' );
+		SP_Config()->update_settings( array( 'host' => 'http://asdftestblog1.files.wordpress.com' ) );
+		$beat_result = SP_Heartbeat()->check_beat( true );
+		SP_Config()->update_settings( array( 'host' => $host ) );
+
+		$this->assertFalse( $beat_result );
+		$this->assertLessThan( $was_scheduled, wp_next_scheduled( 'sp_heartbeat' ) );
+	}
+
+	function test_heartbeat_decreases_frequency_after_error() {
```

Comment:
to avoid confusion with `test_heartbeat_increases_frequency_after_error()` maybe this test could be `test_heartbeat_decreases_frequency_after_error_resolved()`